File: python/tasks/sprite.py
"""
Refactored Sprite Task - "Dumber" Python focused on sprite sheet generation only.

This version:
- Receives explicit S3 paths and sprite parameters from Elixir
- Focuses only on sprite sheet generation using FFmpeg
- Returns structured data about sprite artifacts instead of managing database state
- No database connections or state management
"""

import argparse
import json
import logging
import math
import os
import re
import shutil
import subprocess
import tempfile
import threading
from datetime import datetime
from pathlib import Path

import boto3
from botocore.exceptions import ClientError, NoCredentialsError
from utils import sanitize_filename
from utils.process_utils import run_ffmpeg_command, run_ffprobe_command

# --- Logging Configuration ---
logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger(__name__)

# --- Task Configuration ---
SPRITE_TILE_WIDTH = 480
SPRITE_TILE_HEIGHT = -1  # -1 preserves aspect ratio
SPRITE_FPS = 24
SPRITE_COLS = 5
JPEG_QUALITY = 3  # FFmpeg qscale:v value (lower = better quality)

class S3TransferProgress:
    """Callback for boto3 reporting progress via logger"""
    
    def __init__(self, total_size, filename, logger_instance, throttle_percentage=25):
        self._filename = filename
        self._total_size = total_size
        self._seen_so_far = 0
        self._lock = threading.Lock()
        self._logger = logger_instance
        self._throttle_percentage = max(1, min(int(throttle_percentage), 100))
        self._last_logged_percentage = -1

    def __call__(self, bytes_amount):
        try:
            with self._lock:
                self._seen_so_far += bytes_amount
                current_percentage = 0
                if self._total_size > 0:
                    current_percentage = int((self._seen_so_far / self._total_size) * 100)
                
                should_log = (
                    current_percentage >= self._last_logged_percentage + self._throttle_percentage
                    and current_percentage < 100
                ) or (current_percentage == 100 and self._last_logged_percentage != 100)

                if should_log:
                    size_mb = self._seen_so_far / (1024 * 1024)
                    total_size_mb = self._total_size / (1024 * 1024)
                    self._logger.info(
                        f"S3 Upload: {self._filename} - {current_percentage}% complete "
                        f"({size_mb:.2f}/{total_size_mb:.2f} MiB)"
                    )
                    self._last_logged_percentage = current_percentage
        except Exception as e:
            self._logger.error(f"S3 Progress error: {e}")





def get_video_metadata(video_path: str) -> dict:
    """Extracts detailed video metadata using ffprobe."""
    logger.info(f"Probing video metadata: {video_path}")
    
    ffprobe_cmd = [
        "-v", "error", "-select_streams", "v:0",
        "-show_entries", "stream=duration,r_frame_rate,nb_frames",
        "-count_frames", "-of", "json", str(video_path)
    ]
    
    stdout, _ = run_ffprobe_command(ffprobe_cmd, "FFprobe for video metadata")
    
    try:
        probe_data_list = json.loads(stdout).get('streams', [])
        if not probe_data_list:
            raise ValueError(f"ffprobe found no video streams in {video_path}")
        
        probe_data = probe_data_list[0]
        logger.debug(f"ffprobe result: {probe_data}")

        # Extract duration
        duration_str = probe_data.get('duration')
        if duration_str:
            duration = float(duration_str)
        else:
            raise ValueError("Could not determine video duration")

        # Extract frame rate
        fps_str = probe_data.get('r_frame_rate', '0/1')
        if '/' in fps_str:
            num_str, den_str = fps_str.split('/')
            num, den = int(num_str), int(den_str)
            if den > 0:
                fps = num / den
            else:
                raise ValueError("Invalid frame rate denominator")
        else:
            fps = float(fps_str)

        # Extract total frames
        nb_frames_str = probe_data.get('nb_frames')
        if nb_frames_str:
            total_frames = int(nb_frames_str)
            logger.info(f"Using ffprobe nb_frames: {total_frames}")
        else:
            # Calculate from duration and fps
            total_frames = math.ceil(duration * fps)
            logger.info(f"Calculated total frames from duration*fps: {total_frames}")

        if duration <= 0 or fps <= 0 or total_frames <= 0:
            raise ValueError(f"Invalid video metadata: duration={duration:.3f}s, fps={fps:.3f}, frames={total_frames}")
        
        return {
            "duration": duration,
            "fps": fps,
            "total_frames": total_frames
        }

    except (json.JSONDecodeError, KeyError, ValueError, TypeError) as e:
        logger.error(f"Failed to parse video metadata: {e}")
        raise ValueError(f"Video metadata parsing failed: {e}")

def generate_sprite_sheet(
    video_path: str,
    output_path: Path,
    clip_identifier: str,
    sprite_params: dict
) -> dict:
    """
    Generates a sprite sheet from a video file.
    Returns metadata about the generated sprite sheet.
    """
    # Get sprite parameters
    tile_width = sprite_params.get("tile_width", SPRITE_TILE_WIDTH)
    tile_height = sprite_params.get("tile_height", SPRITE_TILE_HEIGHT)
    sprite_fps = sprite_params.get("fps", SPRITE_FPS)
    cols = sprite_params.get("cols", SPRITE_COLS)
    
    # Get video metadata
    metadata = get_video_metadata(str(video_path))
    duration = metadata["duration"]
    video_fps = metadata["fps"]
    
    # Calculate sprite parameters using video's native FPS
    effective_sprite_fps = min(video_fps, sprite_fps)  # Don't exceed video's native FPS
    num_sprite_frames = math.ceil(duration * effective_sprite_fps)
    
    if num_sprite_frames <= 0:
        raise ValueError(f"Video too short for sprite generation: {duration:.3f}s")
    
    # Generate output filename
    base_filename = sanitize_filename(f"{clip_identifier}_sprite")
    timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
    output_filename = f"{base_filename}_{effective_sprite_fps}fps_w{tile_width}_c{cols}_{timestamp}.jpg"
    final_output_path = output_path / output_filename
    
    # Calculate grid dimensions
    num_rows = max(1, math.ceil(num_sprite_frames / cols))
    
    # Build FFmpeg video filter
    vf_option = f"fps={effective_sprite_fps},scale={tile_width}:{tile_height}:flags=neighbor,tile={cols}x{num_rows}"
    
    # Generate sprite sheet using FFmpeg
    ffmpeg_cmd = [
        "-y", "-i", str(video_path),
        "-vf", vf_option,
        "-an",  # No audio
        "-qscale:v", str(JPEG_QUALITY),
        str(final_output_path)
    ]
    
    run_ffmpeg_command(["ffmpeg"] + ffmpeg_cmd, f"Generating sprite sheet")
    
    if not final_output_path.exists():
        raise RuntimeError("Sprite sheet generation failed - output file not created")
    
    # Get file size
    file_size = final_output_path.stat().st_size
    
    logger.info(f"Successfully generated sprite sheet: {output_filename} ({file_size} bytes)")
    
    return {
        "local_path": str(final_output_path),
        "filename": output_filename,
        "file_size": file_size,
        "sprite_metadata": {
            "effective_fps": effective_sprite_fps,
            "num_frames": num_sprite_frames,
            "cols": cols,
            "rows": num_rows,
            "tile_width": tile_width,
            "tile_height": tile_height,
            "video_duration": duration,
            "video_fps": video_fps
        }
    }

def run_sprite(
    clip_id: int,
    input_s3_path: str,
    output_s3_prefix: str,
    overwrite: bool = False,
    **kwargs
):
    """
    Downloads clip video, generates sprite sheet, and uploads to S3.
    Returns structured data about the sprite artifact instead of managing database state.
    
    Args:
        clip_id: The ID of the clip (for reference only)
        input_s3_path: S3 path to the clip video file
        output_s3_prefix: S3 prefix where to upload sprite sheet
        overwrite: Whether to overwrite existing sprites (for compatibility)
        **kwargs: Additional sprite parameters
    
    Returns:
        dict: Structured data about the sprite artifact including S3 key and metadata
    """
    logger.info(f"RUNNING SPRITE for clip_id: {clip_id}")
    logger.info(f"Input: {input_s3_path}, Output prefix: {output_s3_prefix}, Overwrite: {overwrite}")
    
    # Extract sprite parameters from kwargs
    sprite_params = {
        "tile_width": kwargs.get("tile_width", SPRITE_TILE_WIDTH),
        "tile_height": kwargs.get("tile_height", SPRITE_TILE_HEIGHT),
        "fps": kwargs.get("sprite_fps", SPRITE_FPS),
        "cols": kwargs.get("cols", SPRITE_COLS)
    }
    
    # Get S3 resources from environment (provided by Elixir)
    s3_bucket_name = os.getenv("S3_BUCKET_NAME")
    if not s3_bucket_name:
        raise ValueError("S3_BUCKET_NAME environment variable not set")

    try:
        s3_client = boto3.client("s3")
    except NoCredentialsError:
        logger.error("S3 credentials not found")
        raise

    # Check dependencies
    if not shutil.which("ffmpeg"):
        raise FileNotFoundError("ffmpeg not found in PATH")
    if not shutil.which("ffprobe"):
        raise FileNotFoundError("ffprobe not found in PATH")

    with tempfile.TemporaryDirectory() as temp_dir:
        temp_dir_path = Path(temp_dir)
        
        try:
            # Step 1: Download clip video from S3
            local_video_path = temp_dir_path / Path(input_s3_path).name
            logger.info(f"Downloading s3://{s3_bucket_name}/{input_s3_path} to {local_video_path}")
            s3_client.download_file(s3_bucket_name, input_s3_path, str(local_video_path))

            # Step 2: Generate sprite sheet
            clip_identifier = f"clip_{clip_id}"
            sprite_result = generate_sprite_sheet(
                str(local_video_path),
                temp_dir_path,
                clip_identifier,
                sprite_params
            )

            # Step 3: Upload sprite sheet to S3
            s3_key = f"{output_s3_prefix}/{sprite_result['filename']}"
            local_path = sprite_result['local_path']
            file_size = sprite_result['file_size']
            
            progress_callback = S3TransferProgress(file_size, sprite_result['filename'], logger)
            
            logger.info(f"Uploading {local_path} to s3://{s3_bucket_name}/{s3_key}")
            s3_client.upload_file(
                local_path,
                s3_bucket_name,
                s3_key,
                Callback=progress_callback
            )

            # Return structured data for Elixir to process
            return {
                "status": "success",
                "artifacts": [{
                    "artifact_type": "sprite_sheet",
                    "s3_key": s3_key,
                    "metadata": {
                        "file_size": file_size,
                        "filename": sprite_result['filename'],
                        **sprite_result['sprite_metadata']
                    }
                }],
                "metadata": {
                    "clip_id": clip_id,
                    "sprite_generated": True
                }
            }

        except Exception as e:
            logger.error(f"Error processing sprite: {e}")
            raise

def main():
    """Main entry point for standalone execution"""
    parser = argparse.ArgumentParser(description="Sprite sheet generation task")
    parser.add_argument("--clip-id", type=int, required=True)
    parser.add_argument("--input-s3-path", required=True)
    parser.add_argument("--output-s3-prefix", required=True)
    parser.add_argument("--overwrite", action="store_true", default=False)
    parser.add_argument("--tile-width", type=int, default=SPRITE_TILE_WIDTH)
    parser.add_argument("--sprite-fps", type=int, default=SPRITE_FPS)
    parser.add_argument("--cols", type=int, default=SPRITE_COLS)
    
    args = parser.parse_args()
    
    try:
        result = run_sprite(
            args.clip_id,
            args.input_s3_path,
            args.output_s3_prefix,
            args.overwrite,
            tile_width=args.tile_width,
            sprite_fps=args.sprite_fps,
            cols=args.cols
        )
        print(json.dumps(result, indent=2))
    except Exception as e:
        error_result = {
            "status": "error",
            "error": str(e),
            "error_type": type(e).__name__
        }
        print(json.dumps(error_result, indent=2))
        exit(1)

if __name__ == "__main__":
    main() 

File: lib/heaters/clip/transform/sprite.ex
defmodule Heaters.Clip.Transform.Sprite do
  @moduledoc """
  Sprite sheet generation operations using FFmpeg via ffmpex.

  This module replicates the functionality from sprite.py, handling:
  - Downloading clip videos from S3
  - Extracting video metadata using ffprobe
  - Generating sprite sheets using ffmpeg
  - Uploading sprite artifacts to S3
  - Managing clip state transitions
  """

  import FFmpex
  use FFmpex.Options

  alias Heaters.Clips.Clip
  alias Heaters.Clip.Queries, as: ClipQueries
  alias Heaters.Utils
  require Logger

  # Configuration constants matching sprite.py
  @sprite_tile_width 480
  @sprite_tile_height -1  # -1 preserves aspect ratio
  @sprite_fps 24
  @sprite_cols 5
  @jpeg_quality 3  # FFmpeg qscale:v value (lower = better quality)

  @type sprite_params :: %{
    tile_width: integer(),
    tile_height: integer(),
    fps: integer(),
    cols: integer()
  }

  @type sprite_result :: %{
    status: String.t(),
    artifacts: list(map()),
    metadata: map()
  }

  @doc """
  Main entry point for generating a sprite sheet for a clip.

  This function orchestrates the entire sprite generation process:
  1. Downloads the clip video from S3
  2. Extracts video metadata using ffprobe
  3. Generates a sprite sheet using ffmpeg
  4. Uploads the sprite sheet to S3
  5. Returns structured data about the sprite artifact

  Args:
    clip_id: The ID of the clip for sprite generation
    sprite_params: Optional sprite generation parameters

  Returns:
    {:ok, result} on success with sprite artifact data
    {:error, reason} on failure
  """
  @spec run_sprite(integer(), map()) :: {:ok, sprite_result()} | {:error, any()}
  def run_sprite(clip_id, sprite_params \\ %{}) do
    Logger.info("Sprite: Starting sprite generation for clip_id: #{clip_id}")

    with {:ok, temp_dir} <- create_temp_directory() do
      try do
        with {:ok, clip} <- ClipQueries.get_clip_with_artifacts(clip_id),
             {:ok, final_sprite_params} <- build_sprite_params(sprite_params),
             {:ok, local_video_path} <- download_clip_video(clip, temp_dir),
             {:ok, video_metadata} <- get_video_metadata(local_video_path),
             {:ok, sprite_data} <- generate_sprite_sheet(local_video_path, temp_dir, clip_id, final_sprite_params, video_metadata),
             {:ok, uploaded_sprite_data} <- upload_sprite_sheet(sprite_data, clip) do

          result = %{
            status: "success",
            artifacts: [%{
              artifact_type: "sprite_sheet",
              s3_key: uploaded_sprite_data.s3_key,
              metadata: uploaded_sprite_data.metadata
            }],
            metadata: %{
              clip_id: clip_id,
              sprite_generated: true
            }
          }

          Logger.info("Sprite: Successfully completed sprite generation for clip_id: #{clip_id}")
          {:ok, result}
        else
          error ->
            Logger.error("Sprite: Failed to generate sprite for clip_id: #{clip_id}, error: #{inspect(error)}")
            error
        end
      after
        Logger.info("Sprite: Cleaning up temp directory: #{temp_dir}")
        File.rm_rf(temp_dir)
      end
    else
      error ->
        Logger.error("Sprite: Could not create temp directory: #{inspect(error)}")
        error
    end
  end

  # Private functions implementing the core sprite logic

  @spec build_sprite_params(map()) :: {:ok, sprite_params()} | {:error, any()}
  defp build_sprite_params(input_params) do
    sprite_params = %{
      tile_width: Map.get(input_params, :tile_width, @sprite_tile_width),
      tile_height: Map.get(input_params, :tile_height, @sprite_tile_height),
      fps: Map.get(input_params, :sprite_fps, @sprite_fps),
      cols: Map.get(input_params, :cols, @sprite_cols)
    }

    {:ok, sprite_params}
  end

  @spec create_temp_directory() :: {:ok, String.t()} | {:error, any()}
  defp create_temp_directory do
    case System.tmp_dir() do
      nil -> {:error, "Could not access system temp directory"}
      temp_root ->
        temp_dir = Path.join(temp_root, "heaters_sprite_#{System.unique_integer([:positive])}")
        case File.mkdir_p(temp_dir) do
          :ok -> {:ok, temp_dir}
          {:error, reason} -> {:error, "Failed to create temp directory: #{reason}"}
        end
    end
  end

  @spec download_clip_video(Clip.t(), String.t()) :: {:ok, String.t()} | {:error, any()}
  defp download_clip_video(%Clip{clip_filepath: s3_path} = _clip, temp_dir) do
    local_filename = Path.basename(s3_path)
    local_path = Path.join(temp_dir, local_filename)

    Logger.info("Sprite: Downloading #{s3_path} to #{local_path}")

    bucket_name = get_s3_bucket_name()
    s3_key = String.trim_leading(s3_path, "/")

    try do
      # Use streaming download for better memory efficiency with large files
      file = File.open!(local_path, [:write, :binary])

      try do
        ExAws.S3.get_object(bucket_name, s3_key)
        |> ExAws.stream!()
        |> Enum.each(&IO.binwrite(file, &1))
      after
        File.close(file)
      end

      if File.exists?(local_path) do
        {:ok, local_path}
      else
        {:error, "Downloaded file does not exist at #{local_path}"}
      end
    rescue
      error ->
        Logger.error("Sprite: Failed to stream download from S3: #{Exception.message(error)}")
        {:error, "Failed to download from S3: #{Exception.message(error)}"}
    end
  end

  @spec get_video_metadata(String.t()) :: {:ok, map()} | {:error, any()}
  defp get_video_metadata(video_path) do
    Logger.info("Sprite: Probing video metadata: #{video_path}")

    try do
      # Use System.cmd to call ffprobe directly to get detailed metadata
      # Replicating: ffprobe -v error -select_streams v:0 -show_entries stream=duration,r_frame_rate,nb_frames -count_frames -of json video_path
      {output, exit_code} = System.cmd("ffprobe", [
        "-v", "error",
        "-select_streams", "v:0",
        "-show_entries", "stream=duration,r_frame_rate,nb_frames",
        "-count_frames",
        "-of", "json",
        video_path
      ])

      if exit_code == 0 do
        case Jason.decode(output) do
          {:ok, %{"streams" => [probe_data | _]}} ->
            parse_video_metadata(probe_data, video_path)
          {:ok, %{"streams" => []}} ->
            {:error, "ffprobe found no video streams in #{video_path}"}
          {:error, reason} ->
            {:error, "Failed to parse ffprobe JSON output: #{inspect(reason)}"}
        end
      else
        Logger.error("Sprite: FFprobe command failed with exit code #{exit_code}")
        {:error, "FFprobe command failed with exit code #{exit_code}"}
      end
    rescue
      e ->
        Logger.error("Sprite: Exception getting video metadata: #{Exception.message(e)}")
        {:error, "Exception getting video metadata: #{Exception.message(e)}"}
    end
  end

  @spec parse_video_metadata(map(), String.t()) :: {:ok, map()} | {:error, any()}
  defp parse_video_metadata(probe_data, _video_path) do
    Logger.debug("Sprite: ffprobe result: #{inspect(probe_data)}")

    with {:ok, duration} <- extract_duration(probe_data),
         {:ok, fps} <- extract_fps(probe_data),
         {:ok, total_frames} <- extract_total_frames(probe_data, duration, fps) do

      if duration <= 0 or fps <= 0 or total_frames <= 0 do
        {:error, "Invalid video metadata: duration=#{duration}s, fps=#{fps}, frames=#{total_frames}"}
      else
        metadata = %{
          duration: duration,
          fps: fps,
          total_frames: total_frames
        }
        {:ok, metadata}
      end
    else
      error -> error
    end
  end

  defp extract_duration(%{"duration" => duration_str}) when is_binary(duration_str) do
    case Float.parse(duration_str) do
      {duration, ""} -> {:ok, duration}
      _ -> {:error, "Could not parse duration: #{duration_str}"}
    end
  end
  defp extract_duration(_), do: {:error, "Could not determine video duration"}

  defp extract_fps(%{"r_frame_rate" => fps_str}) when is_binary(fps_str) do
    case String.split(fps_str, "/") do
      [num_str, den_str] ->
        with {num, ""} <- Integer.parse(num_str),
             {den, ""} <- Integer.parse(den_str),
             true <- den > 0 do
          {:ok, num / den}
        else
          _ -> {:error, "Invalid frame rate: #{fps_str}"}
        end
      _ ->
        case Float.parse(fps_str) do
          {fps, ""} -> {:ok, fps}
          _ -> {:error, "Could not parse frame rate: #{fps_str}"}
        end
    end
  end
  defp extract_fps(_), do: {:error, "Could not determine frame rate"}

  defp extract_total_frames(%{"nb_frames" => nb_frames_str}, _duration, _fps) when is_binary(nb_frames_str) do
    case Integer.parse(nb_frames_str) do
      {total_frames, ""} when total_frames > 0 ->
        Logger.info("Sprite: Using ffprobe nb_frames: #{total_frames}")
        {:ok, total_frames}
      _ ->
        {:error, "Invalid nb_frames: #{nb_frames_str}"}
    end
  end
  defp extract_total_frames(_, duration, fps) do
    # Calculate from duration and fps as fallback
    total_frames = ceil(duration * fps)
    Logger.info("Sprite: Calculated total frames from duration*fps: #{total_frames}")
    {:ok, total_frames}
  end

  @spec generate_sprite_sheet(String.t(), String.t(), integer(), sprite_params(), map()) :: {:ok, map()} | {:error, any()}
  defp generate_sprite_sheet(video_path, output_dir, clip_id, sprite_params, video_metadata) do
    %{
      tile_width: tile_width,
      tile_height: tile_height,
      fps: sprite_fps,
      cols: cols
    } = sprite_params

    %{
      duration: duration,
      fps: video_fps,
      total_frames: _total_frames
    } = video_metadata

    # Calculate sprite parameters using video's native FPS
    effective_sprite_fps = min(video_fps, sprite_fps)  # Don't exceed video's native FPS
    num_sprite_frames = ceil(duration * effective_sprite_fps)

    if num_sprite_frames <= 0 do
      {:error, "Video too short for sprite generation: #{duration}s"}
    else
      # Generate output filename
      clip_identifier = "clip_#{clip_id}"
      base_filename = Utils.sanitize_filename("#{clip_identifier}_sprite")
      timestamp = DateTime.utc_now() |> DateTime.to_unix() |> Integer.to_string()
      output_filename = "#{base_filename}_#{trunc(effective_sprite_fps)}fps_w#{tile_width}_c#{cols}_#{timestamp}.jpg"
      final_output_path = Path.join(output_dir, output_filename)

      # Calculate grid dimensions
      num_rows = max(1, ceil(num_sprite_frames / cols))

      Logger.info("Sprite: Generating sprite sheet: #{output_filename}")
      Logger.info("Sprite: Parameters - fps: #{effective_sprite_fps}, frames: #{num_sprite_frames}, grid: #{cols}x#{num_rows}")

      case create_sprite_with_ffmpeg(video_path, final_output_path, effective_sprite_fps, tile_width, tile_height, cols, num_rows) do
        {:ok, file_size} ->
          sprite_data = %{
            local_path: final_output_path,
            filename: output_filename,
            file_size: file_size,
            sprite_metadata: %{
              effective_fps: effective_sprite_fps,
              num_frames: num_sprite_frames,
              cols: cols,
              rows: num_rows,
              tile_width: tile_width,
              tile_height: tile_height,
              video_duration: duration,
              video_fps: video_fps
            }
          }
          {:ok, sprite_data}
        error -> error
      end
    end
  end

  @spec create_sprite_with_ffmpeg(String.t(), String.t(), float(), integer(), integer(), integer(), integer()) :: {:ok, integer()} | {:error, any()}
  defp create_sprite_with_ffmpeg(input_path, output_path, sprite_fps, tile_width, tile_height, cols, num_rows) do
    try do
      # Build FFmpeg video filter matching the Python implementation:
      # vf_option = f"fps={effective_sprite_fps},scale={tile_width}:{tile_height}:flags=neighbor,tile={cols}x{num_rows}"
      vf_filter = "fps=#{sprite_fps},scale=#{tile_width}:#{tile_height}:flags=neighbor,tile=#{cols}x#{num_rows}"

      # Create FFmpeg command matching the Python implementation exactly:
      # ffmpeg -y -i input -vf "fps=...,scale=...,tile=..." -an -qscale:v 3 output.jpg
      command =
        FFmpex.new_command()
        |> add_global_option(option_y())  # Overwrite output file
        |> add_input_file(input_path)
        |> add_output_file(output_path)
        |> add_file_option(option_vf(vf_filter))
        |> add_file_option(option_an())  # No audio
        |> add_file_option(option_qscale("#{@jpeg_quality}:v"))

      case FFmpex.execute(command) do
        {:ok, _output} ->
          # Verify the file was created and get its size
          case File.stat(output_path) do
            {:ok, %File.Stat{size: file_size}} ->
              Logger.info("Sprite: Successfully generated sprite sheet: #{Path.basename(output_path)} (#{file_size} bytes)")
              {:ok, file_size}
            {:error, reason} ->
              {:error, "Created sprite file not found: #{reason}"}
          end
        {:error, reason} ->
          Logger.error("Sprite: FFmpeg failed: #{inspect(reason)}")
          {:error, "FFmpeg sprite generation failed: #{inspect(reason)}"}
      end
    rescue
      e ->
        {:error, "Exception creating sprite sheet: #{inspect(e)}"}
    end
  end

  @spec upload_sprite_sheet(map(), Clip.t()) :: {:ok, map()} | {:error, any()}
  defp upload_sprite_sheet(sprite_data, %Clip{source_video_id: source_video_id, id: clip_id}) do
    bucket_name = get_s3_bucket_name()
    output_prefix = "source_videos/#{source_video_id}/clips/#{clip_id}/sprites"
    s3_key = "#{output_prefix}/#{sprite_data.filename}"

    Logger.info("Sprite: Uploading sprite sheet to s3://#{bucket_name}/#{s3_key}")

    case ExAws.S3.Upload.stream_file(sprite_data.local_path)
         |> ExAws.S3.upload(bucket_name, s3_key)
         |> ExAws.request() do
      {:ok, _result} ->
        uploaded_sprite_data = %{
          s3_key: s3_key,
          metadata: %{
            file_size: sprite_data.file_size,
            filename: sprite_data.filename
          } |> Map.merge(sprite_data.sprite_metadata)
        }
        {:ok, uploaded_sprite_data}
      {:error, reason} ->
        Logger.error("Sprite: Failed to upload sprite sheet: #{inspect(reason)}")
        {:error, "Failed to upload sprite sheet: #{inspect(reason)}"}
    end
  end

  @spec get_s3_bucket_name() :: String.t()
  defp get_s3_bucket_name do
    case Application.get_env(:heaters, :s3_bucket) do
      nil ->
        raise "S3 bucket name not configured. Please set :s3_bucket in application config."
      bucket_name ->
        bucket_name
    end
  end
end
